core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -378,6 +384,26 @@ class ReplicaFetcherThread(name: String,
     result
   }
 
+  /*
+    Decides whether or not we should send a LeaderEpochRequest
+      by checking the broker message format and all topics' message format versions
+   */
+  private def shouldSendLeaderEpochRequest(partitions: Map[TopicPartition, Int]): Boolean = {
+    if (!brokerSupportsLeaderEpochRequest)
+      return false
+
+    // if all topics' message format do not support epoch requests, skip it
+    partitions.map { case (tp, _) => topicSupportsEpochRequest(tp) }.reduce(_ || _)
```

Comment:
If we are going to skip partitions on older message format, we could do that in `buildLeaderEpochRequest()` function, where we already skip partitions with no recorded leader epoch or partitions without epoch cache. I think you are right about the need to check message format, and possibly not requiring to check inter-broker protocol in that case, but maybe @junrao  could confirm.